op-node/rollup/driver/sequencer.go

```diff
@@ -0,0 +1,101 @@
+	d.Log.Info("creating new block", "parent", l2Head, "l1Origin", l1Origin)
+	if d.buildingID != (eth.PayloadID{}) { // This may happen when we decide to build a different block in response to a reorg. Or when previous block building failed.
+		d.Log.Warn("did not finish previous block building, starting new building now", "prev_onto", d.buildingOnto.HeadBlockHash, "prev_payload_id", d.buildingID, "new_onto", l2Head)
+	}
+
+	fetchCtx, cancel := context.WithTimeout(ctx, time.Second*20)
+	defer cancel()
+
+	attrs, err := derive.PreparePayloadAttributes(fetchCtx, d.Config, d.L1, l2Head, l2Head.Time+d.Config.BlockTime, l1Origin.ID())
+	if err != nil {
+		return err
+	}
+
+	// If our next L2 block timestamp is beyond the Sequencer drift threshold, then we must produce
+	// empty blocks (other than the L1 info deposit and any user deposits). We handle this by
+	// setting NoTxPool to true, which will cause the Sequencer to not include any transactions
+	// from the transaction pool.
+	attrs.NoTxPool = uint64(attrs.Timestamp) >= l1Origin.Time+d.Config.MaxSequencerDrift
+
+	// And construct our fork choice state. This is our current fork choice state and will be
+	// updated as a result of executing the block based on the attributes described above.
+	fc := eth.ForkchoiceState{
+		HeadBlockHash:      l2Head.Hash,
+		SafeBlockHash:      l2SafeHead.Hash,
+		FinalizedBlockHash: l2Finalized.Hash,
+	}
+	// Start a payload building process.
+	id, errTyp, err := derive.StartPayload(ctx, d.L2, fc, attrs)
+	if err != nil {
+		return fmt.Errorf("failed to start building on top of L2 chain %s, error (%d): %w", l2Head, errTyp, err)
+	}
+	d.buildingOnto = fc
+	d.buildingID = id
+	return nil
+}
+
+// CompleteBuildingBlock takes the current block that is being built, and asks the engine to complete the building, seal the block, and persist it as canonical.
+func (d *Sequencer) CompleteBuildingBlock(ctx context.Context) (*eth.ExecutionPayload, error) {
+	if d.buildingID == (eth.PayloadID{}) {
+		return nil, fmt.Errorf("cannot complete payload building: not currently building a payload")
+	}
+
+	// Actually execute the block and add it to the head of the chain.
+	payload, errTyp, err := derive.ConfirmPayload(ctx, d.Log, d.L2, d.buildingOnto, d.buildingID, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to complete building on top of L2 chain %s, error (%d): %w", d.buildingOnto.HeadBlockHash, errTyp, err)
+	}
+	return payload, nil
+}
+
+// createNewBlock sequences a L2 block with immediate building and sealing.
+// Deprecated: the sequencer should build the block in two steps, adjusted to maximize building time, within the block-time span, or faster when catching up.
+func (d *Sequencer) createNewBlock(ctx context.Context, l2Head eth.L2BlockRef, l2SafeHead eth.BlockID, l2Finalized eth.BlockID, l1Origin eth.L1BlockRef) (eth.L2BlockRef, *eth.ExecutionPayload, error) {
```

Comment:
For this file, I'm fine leaving the function signatures as is, but I'd suggest that the `start/stop building` functions to be private and this function to be public with the deprecated warning removed. The reason for this is that the `CompleteBuildingBlock` function reuses the forkchoice update struct. This means that if there are other updates to `safeHead` or `finalizedHead` the forkchoice update in `completeBuildingBlock` will be out of date. To avoid this the `Completion` API could be changed or it needs to be noted that new modifications to the l2 execution engine state should be made in between the `start` and `complete` calls.

Reply:
Will add a note about the forkchoice state usage and consistency between start and complete. I do need to make the start/complete functions public so I can call into them from the op-e2e action testing package, or we end up with duplicate code in the action-testing that doesn't cover the actual driver code.